tests/test_options.py

```diff
@@ -1455,31 +1688,252 @@ def cli(one, two):
+            [],
+            True,
+        ),
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": True},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            ["--opt1"],
+            Class1,
+        ),
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": True},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            ["--opt2"],
+            Class2,
+        ),
+        # Setting the default to a class, an instance of the class is returned instead
+        # of the class itself, because the default is allowed to be callable (and
+        # consummd). And this what happens the type is.
+        (
+            {"flag_value": Class1, "default": Class1},
+            {"flag_value": Class2},
+            [],
+            re.compile(r"'<test_options.Class1 object at 0x[0-9A-Fa-f]+>'"),
+        ),
+        (
+            {"flag_value": Class1, "default": Class2},
+            {"flag_value": Class2},
+            [],
+            re.compile(r"'<test_options.Class2 object at 0x[0-9A-Fa-f]+>'"),
+        ),
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": Class1},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            [],
+            re.compile(r"<test_options.Class1 object at 0x[0-9A-Fa-f]+>"),
+        ),
+        (
+            {"flag_value": Class1, "type": UNPROCESSED, "default": Class2},
+            {"flag_value": Class2, "type": UNPROCESSED},
+            [],
+            re.compile(r"<test_options.Class2 object at 0x[0-9A-Fa-f]+>"),
+        ),
+        ###
+        ### Reproduces https://github.com/pallets/click/issues/2012#issuecomment-892437060
+        ###
+        (
+            {"flag_value": 1, "default": True},
+            {"flag_value": "1"},
+            [],
+            1,
+        ),
+        (
+            {"flag_value": 1, "type": int, "default": True},
+            {"flag_value": "1", "type": int},
+            [],
+            1,
         ),
     ],
 )
-def test_default_type_error_raises(multiple, nargs, default, expected_message):
-    with pytest.raises(ValueError, match=expected_message):
-        click.Option(
-            ["--test"],
-            multiple=multiple,
-            nargs=nargs,
-            default=default,
-        )
+def test_custom_type_flag_value_dual_options(
+    runner, opt1_params, opt2_params, args, expected
+):
+    @click.command()
+    @click.option("--opt1", "dual_option", **opt1_params)
+    @click.option("--opt2", "dual_option", **opt2_params)
+    def cli(dual_option):
+        click.echo(repr(dual_option), nl=False)
+
+    result = runner.invoke(cli, args)
+    if isinstance(expected, re.Pattern):
+        assert re.match(expected, result.output)
+    else:
+        assert result.output == repr(expected)
+
+
+def test_custom_type_frozenset_flag_value(runner):
+    """A reproduction of https://github.com/pallets/click/issues/2610"""
```

Comment:
Describe what's being tested, not only a link.

Reply:
Done: https://github.com/pallets/click/pull/3030/commits/472bf383bcf403487b939242ddcd0bbb54257bb3